apps/prairielearn/src/ee/lib/billing/plan-grants.ts

```diff
@@ -46,3 +57,45 @@ export async function checkPlanGrants(res: Response) {
 
   return planGrantsMatchPlanFeatures(planGrantNames, requiredPlans);
 }
+
+export async function checkPlanGrantsForQuestion(res: Response) {
+  if (userHasRole(res)) {
+    return true;
+  }
+
+  // For the time being, put this being a feature flag so that we can land
+  // this change before we've created plan grants for all existing institutions.
+  const shouldCheck = await features.enabledFromLocals(
+    'enforce-plan-grants-for-questions',
+    res.locals,
+  );
+  if (!shouldCheck) {
+    return true;
+  }
+
+  const question = QuestionSchema.parse(res.locals.question);
+  const requiredFeatures: PlanFeatureName[] = [];
+  if (question.external_grading_enabled) {
+    requiredFeatures.push('external-grading');
+  }
+  if (question.workspace_image) {
+    requiredFeatures.push('workspaces');
+  }
+
+  // If there aren't any features to check, don't even bother hitting the database.
+  if (requiredFeatures.length === 0) {
+    return true;
+  }
```

Comment:
Re-order before feature flag check so that we can avoid an unnecessary DB query.

Reply:
Done in 68dbce3.